### PR TITLE
Use aiosmtpd for the smtp.Server class

### LINKED ISSUE #12: Module smtpd marked deprecated in Python 3.10
Apparently the SMTP module used in `pytest_loclaserver/smtp.py` will be deprecated in Python version 3.10. 

```
.tox/py310/lib/python3.10/site-packages/pytest_localserver/smtp.py:12
  /home/runner/work/pytest-localserver/pytest-localserver/.tox/py310/lib/python3.10/site-packages/pytest_localserver/smtp.py:12: DeprecationWarning: The smtpd module is deprecated and unmaintained.  Please see aiosmtpd (https://aiosmtpd.readthedocs.io/) for the recommended replacement.
    import smtpd
```
https://github.com/redtoad/pytest-localserver/runs/4191982788?check_suite_focus=true#step:5:218

Suggested replacement:
https://aiosmtpd.readthedocs.io/

### PR DESCRIPTION
**Pull request by diazona**
I made these changes to switch from using Python's standard [smtpd](https://docs.python.org/3/library/smtpd.html) module, which is deprecated, to the external [aiosmtpd](https://aiosmtpd.readthedocs.io/en/latest/) package. The changes are somewhat substantial, but most of them are just meant to keep the Server class interface backward compatible and to support old versions of Python and smtpd. Once we drop support for older versions, it will simplify the implementation quite a bit.

One thing to note: it's not entirely clear how actively aiosmtpd is maintained, since the last commit on Github was over a year ago and there's a featured issue asking for help with the package. So I can imagine a future where the suggested replacement for smtpd changes to something other than aiosmtpd. It's _possible_ we might want to defer this merge request until the status of aiosmtpd as a replacement for smtpd becomes more clear.

Closes #12

### CODE REVIEW

**Comment by diazona** [comment]
@redtoad @coordt any input on this?

**Comment by diazona** [comment]
@nicoddemus if you don't mind, I'd appreciate your thoughts about introducing a dependency on `aiosmtpd` given its unclear maintenance status. Maybe you know more than I do about what the future of that package looks like :shrug: (Of course any other feedback on the PR is welcome too!)

**Comment by nicoddemus** [comment]
Hi @diazona,

Thanks a lot for the work on this! I didn't review it in depth, but the code looks good and well documented.

> Maybe you know more than I do about what the future of that package looks like

Unfortunately I don't, AFAIK @coordt and @redtoad are still the maintainers, might have been lack of time to go over the PR yet.

If you are willing to maintain a plugin, and given that it is possible to have a different/incompatible API that fits the async model better, to have this as a separate plugin altogether: if users need to port the code, might make sense to migrate the plugin as well.

**Comment by diazona** [comment]
@nicoddemus Thanks for the review!

Sorry if I'm misunderstanding what you meant, but it sounds like you might have missed that I'm also a maintainer of this package (`pytest-localserver`), so I'll be able to merge the PR myself. I just really wanted to get at least one extra set of eyes on this before merging it, and since it's been two weeks without hearing from @coordt or @redtoad, I asked you.

The main thing I wanted a second opinion on is whether it makes sense to introduce this dependency on `aiosmtpd`, given that its maintenance status is unclear. I guess if you don't have any special insight on that, I'll try reaching out to the community around that package directly and see what they say.

**Comment by nicoddemus** [comment]
>  but it sounds like you might have missed that I'm also a maintainer of this package (pytest-localserver), so I'll be able to merge the PR myself.

Ahh indeed I missed that, sorry about that!

> The main thing I wanted a second opinion on is whether it makes sense to introduce this dependency on aiosmtpd, given that its maintenance status is unclear.

Ahh sorry I don't have any insight on that. Reaching them sounds good. 👍 

Sorry about the confusion on my part!

**Review by coordt — APPROVED** [review]

**Pull request merged**
